lib/improver/spotdata/common_functions.py

```diff
@@ -0,0 +1,408 @@
+
+
+def construct_neighbour_hash(neighbour_finding):
+    """
+    Constructs a hash from the various neighbour finding options. This is used
+    to avoid repeating the same neighbour search more than once.
+
+    Args:
+    -----
+    neighbour_finding : dict
+        A dictionary containing the method, vertical_bias, and land_constraint
+        options for neighbour finding.
+
+    Returns:
+    --------
+    neighbour_hash : string
+        A concatenated string of the options
+        e.g. 'fast_nearest_neighbour-None-False'
+
+    """
+    return '{}-{}-{}'.format(neighbour_finding['method'],
+                             neighbour_finding['vertical_bias'],
+                             neighbour_finding['land_constraint'])
+
+
+def apply_bias(vertical_bias, dzs):
+    """
+    Bias neighbour selection to look for grid points with an
+    altitude that is above or below the site if vertical_bias is
+    not None.
+
+    Args:
+    -----
+    vertical_bias : string/None
+        Sets the preferred vertical displacement of the grid point
+        relative to the site; above/below/None.
+
+    dzs : numpy.array
+        Array of vertical displacments calculated as the subtraction of grid
+        orography altitudes from spot site altitudes.
+
+    Returns:
+    --------
+    dz_subset : numpy.array
+        Indices of grid points that satisfy bias condition if any are
+        available, otherwise it returns the whole set.
+
+    """
+    if vertical_bias == 'above':
+        dz_subset, = np.where(dzs <= 0)
+    elif vertical_bias == 'below':
+        dz_subset, = np.where(dzs >= 0)
+
+    if (vertical_bias is None or len(dz_subset) == 0
+            or len(dz_subset) == len(dzs)):
+        dz_subset = np.arange(len(dzs))
+
+    return dz_subset
+
+
+def xy_test(cube):
+    """
+    Test whether a diagnostic cube is on a latitude/longitude grid or uses an
+    alternative projection.
+
+    Args:
+    -----
+    cube : iris.cube.Cube
+        A diagnostic cube to examine for coordinate system.
+
+    Returns:
+    --------
+    trg_crs : cartopy.crs/None
+        Coordinate system of the diagnostic cube in a cartopy format unless it
+        is already a latitude/longitude grid, in which case None is returned.
+
+    """
+    trg_crs = None
+    if (not cube.coord(axis='x').name() == 'longitude' or
+            not cube.coord(axis='y').name() == 'latitude'):
+        trg_crs = cube.coord_system().as_cartopy_crs()
+    return trg_crs
+
+
+def xy_transform(trg_crs, latitude, longitude):
+    """
+    Transforms latitude/longitude coordinate pairs from a latitude/longitude
+    grid into an alternative projection defined by trg_crs.
+
+    Args:
+    -----
+    trg_crs : cartopy.crs/None
+        Target coordinate system in cartopy format or None.
+
+    latitude : float
+        Latitude coordinate.
+
+    longitude : float
+        Longitude coordinate.
+
+    Returns:
+    --------
+    x, y : floats
```

Comment:
No, but it converts the lat/lons into pure x/y coordinates in the target coordinate system. I think this one actually helps make clear what's going on.